crates/fluvio-benchmark/src/benchmark_config/benchmark_matrix.rs

```diff
@@ -0,0 +1,189 @@
+    type Item = BenchmarkSettings;
+
+    type IntoIter = <Vec<BenchmarkSettings> as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.generate_settings().into_iter()
+    }
+}
+
+impl BenchmarkMatrix {
+    // Impl note: This does allocate for all of the benchmark settings at once, however it made for simpler code
+    // and as there is a very low practical limit for the number of benchmarks that can be run in a reasonable time period, its not an issue that it alloates.
+    // TODO split into smaller chunks
+
+    fn generate_settings(&self) -> Vec<BenchmarkSettings> {
+        let builder = vec![BenchmarkBuilder::new(&self.shared_settings)];
+        builder
+            .cross_iterate(&self.num_records_per_producer_worker_per_batch, |v, b| {
+                b.num_records_per_producer_worker_per_batch = Some(v);
+            })
+            .cross_iterate(&self.producer_batch_size, |v, b| {
+                b.producer_batch_size = Some(v);
+            })
+            .cross_iterate(&self.producer_queue_size, |v, b| {
+                b.producer_queue_size = Some(v);
+            })
+            .cross_iterate(&self.producer_linger_millis, |v, b| {
+                b.producer_linger = Some(Duration::from_millis(v));
+            })
+            .cross_iterate(&self.producer_server_timeout_millis, |v, b| {
+                b.producer_server_timeout = Some(Duration::from_millis(v));
+            })
+            .cross_iterate(&self.producer_compression, |v, b| {
+                b.producer_compression = Some(v);
+            })
+            .cross_iterate(&self.consumer_max_bytes, |v, b| {
+                b.consumer_max_bytes = Some(v);
+            })
+            .cross_iterate(&self.num_concurrent_producer_workers, |v, b| {
+                b.num_concurrent_producer_workers = Some(v);
+            })
+            .cross_iterate(&self.num_concurrent_consumers_per_partition, |v, b| {
+                b.num_concurrent_consumers_per_partition = Some(v);
+            })
+            .cross_iterate(&self.num_partitions, |v, b| {
+                b.num_partitions = Some(v);
+            })
+            .cross_iterate(&self.record_size, |v, b| {
+                b.record_size_strategy = Some(v);
+            })
+            .cross_iterate(&self.record_key_allocation_strategy, |v, b| {
+                b.record_key_allocation_strategy = Some(v);
+            })
+            .build()
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Copy, Clone, Hash, PartialEq, Eq)]
+pub enum RecordKeyAllocationStrategy {
+    /// RecordKey::NULL
+    NoKey,
+    /// All producer workers will use the same key
+    AllShareSameKey,
+
+    /// Each producer will use the same key for each of their records
+    ProducerWorkerUniqueKey,
+
+    /// Each producer will round robin from 0..N for each record produced
+    RoundRobinKey(u64),
+
+    RandomKey,
+}
+
+pub fn get_default_config() -> Vec<BenchmarkMatrix> {
```

Comment:
why does this need to be public?

Reply:
It's called by main function in `bin/fdm.rs`